### PR TITLE
Issue 743 Fix

### PR DESCRIPTION
**Pull request by marcusking01**
Added a nil check before trying to use a host to execute a query.  This was causing nil pointer dereference exceptions as described at https://github.com/gocql/gocql/issues/743. Also made some optimizations to reduce lock scoping in the TokenAwareHostPolicy

### CODE REVIEW

File: policies.go [hunk 1]
```diff
@@ -263,9 +263,6 @@ type tokenAwareHostPolicy struct {
 }
 
 func (t *tokenAwareHostPolicy) SetPartitioner(partitioner string) {
-	t.mu.Lock()
```
Comment (Zariel) [inline policies.go]: im not sure removing this lock is correct, as its protecting concurrent access to `t.partitioner`

Comment (marcusking01) [inline policies.go]: The only thing that SetPartitioner does is call to the fallback policies Partitioner (which in the case of HostPool and RoundRobin are no-ops)  then calls to the resetTokenRing which is where I moved the lock to.

**Comment by Zariel** [comment]
Ill merge this once I confirm there are no data races around the partitioner on the token aware policy

**Comment by Zariel** [comment]
Thanks for the contribution, ive merged this manually. Sorry for the delay!

**Pull request merged**